File: samples/java/DeliveryDispatch/Server/CustomerGateway/src/main/java/systems/zlink/samples/deliverydispatch/server/customergateway/CustomerActorDirectory.java
package systems.zlink.samples.deliverydispatch.server.customergateway;

import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.HashMap;
import java.util.Map;

public final class CustomerActorDirectory {
    private final Object gate = new Object();
    private final Map<String, CustomerActor> actors = new HashMap<>();
    private final Map<String, String> deliveryCustomers = new HashMap<>();

    public void register(CustomerActor actor) {
        synchronized (gate) {
            actors.put(actor.actorId(), actor);
        }
    }

    public void remove(String actorId) {
        synchronized (gate) {
            actors.remove(actorId);
            deliveryCustomers.values().removeIf(actorId::equals);
        }
    }

    public void subscribe(String customerId, String deliveryId) {
        synchronized (gate) {
            deliveryCustomers.put(deliveryId, customerId);
        }
    }

    public void push(Messages.DeliveryStatusChangedReq status) {
        CustomerActor actor;
        synchronized (gate) {
            String customerId = deliveryCustomers.get(status.deliveryId());
            actor = customerId == null ? null : actors.get(customerId);
        }
        if (actor == null) {
            return;
        }
        actor.push(
                new Messages.DeliveryStatusNotify(
                        status.deliveryId(),
                        status.status(),
                        status.courierId(),
                        status.occurredAt()));
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CustomerGateway/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/customergateway/CustomerActorDirectory.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.customergateway

import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusChangedReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.DeliveryStatusNotify

class CustomerActorDirectory {
    private val gate = Any()
    private val actors = mutableMapOf<String, CustomerActor>()
    private val deliveryCustomers = mutableMapOf<String, String>()

    fun register(actor: CustomerActor) {
        synchronized(gate) { actors[actor.actorId()] = actor }
    }

    fun remove(actorId: String) {
        synchronized(gate) {
            actors.remove(actorId)
            deliveryCustomers.values.removeIf(actorId::equals)
        }
    }

    fun subscribe(customerId: String, deliveryId: String) {
        synchronized(gate) { deliveryCustomers[deliveryId] = customerId }
    }

    fun push(status: DeliveryStatusChangedReq) {
        val actor =
            synchronized(gate) { deliveryCustomers[status.deliveryId]?.let { actors[it] } }
                ?: return
        actor.push(
            DeliveryStatusNotify(
                deliveryId = status.deliveryId,
                status = status.status,
                courierId = status.courierId,
                occurredAt = status.occurredAt,
            )
        )
    }
}
